src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLRenderer.m

```diff
@@ -162,27 +245,35 @@ void MTLRenderer_DrawRect(MTLContext *mtlc, BMTLSDOps * dstOps, jint x, jint y,
     id<MTLTexture> dest = dstOps->pTexture;
     J2dTraceLn5(J2D_TRACE_INFO, "MTLRenderer_DrawRect (x=%d y=%d w=%d h=%d), dst tex=%p", x, y, w, h, dest);
 
-    // TODO: use DrawParallelogram(x, y, w, h, lw=1, lh=1)
-    id<MTLRenderCommandEncoder> mtlEncoder = [mtlc.encoderManager getRenderEncoder:dstOps];
-    if (mtlEncoder == nil)
-        return;
+    // Make sure we have space for 8 more vertices in the batch
+    if ( ((currentIndexInBatch + 8) > VERTEX_BATCH_SIZE) ||
+         (currentBatchNo == (MAX_NO_OF_BATCHES - 1))) {
+
+        // encode the vertex batch
+        MTLRenderer_SubmitVertexBatch(mtlc, dstOps);
+    }
+
+    MTLRenderer_SetPrimitiveType(MTLPrimitiveTypeLine);
+
 
     // Translate each vertex by a fraction so
     // that we hit pixel centers.
-    const int verticesCount = 5;
     float fx = (float)x + 0.2f;
     float fy = (float)y + 0.5f;
     float fw = (float)w;
     float fh = (float)h;
-    struct Vertex vertices[5] = {
-            {{fx, fy}},
-            {{fx + fw, fy}},
-            {{fx + fw, fy + fh}},
-            {{fx, fy + fh}},
-            {{fx, fy}},
-    };
-    [mtlEncoder setVertexBytes:vertices length:sizeof(vertices) atIndex:MeshVertexBuffer];
-    [mtlEncoder drawPrimitives:MTLPrimitiveTypeLineStrip vertexStart:0 vertexCount:verticesCount];
+
+    MTLRenderer_AddVertexToBatch(fx, fy);
+    MTLRenderer_AddVertexToBatch(fx+fw, fy);
+
+    MTLRenderer_AddVertexToBatch(fx+fw, fy);
```

Comment:
Why are there several duplicates here ?

Reply:
Earlier logic used to draw a rectangle with MTLPrimitiveTypeLineStrip. Hence, 5 vertices were specified to  draw 4 sides of the rectangle.
Now, the logic has been changed to use MTLPrimitiveTypeLine. Hence, 4 lines need to be specified separately by specifying 8 vertices. There are duplicates since the lines are connected.

We cannot use MTLPrimitiveTypeLineStrip if we want to batch the subsequent draw calls as it draws an unwanted line between previous rectangle and current rectangle.

Reply:
> We cannot use MTLPrimitiveTypeLineStrip if we want to batch the subsequent draw calls as it draws an unwanted line between previous rectangle and current rectangle.

Even if the alpha will be transparent? Just curious.

Reply:
Yes, there will be a transparent line between previous rectangle and current rectangle.
This PR introduces batching of vertices of successive draw calls of the same primitive. If we keep on adding vertices to a common buffer and finally encode draw operation using MTLPrimitiveTypeLineStrip, all those vertices will be connected. If color is changed in between these successive draw calls, we end the current vertex batch and start a new one.

Reply:
But what about the usage of MTLPrimitiveTypeLineStrip and the transparent color to hide the "unwanted line between previous rectangle and current rectangle"?

Reply:
> OR it is not possible to draw the lines using different colors in one step?
Exactly this is the point.  If the color is changed (alpha change is also treated as color change) in between these successive draw calls, we end the current vertex batch and issue a draw command for accumulated vertices.
This is the reason MTLPrimitiveTypeLineStrip and MTLPrimitiveTriangleStrip cannot be used with our current vertex batching logic.